include/wil/safearrays.h

```diff
@@ -0,0 +1,1084 @@
+        //! @{
+
+        //! Returns the lower bound of the given dimension
+        WI_NODISCARD LONG lbound(UINT dim = 1) const
+        {
+            static_assert(wistd::is_same<void, result>::value, "this method requires exceptions or failfast");
+            LONG result = 0;
+            lbound(dim, &result);
+            return result;
+        }
+        //! Returns the bupper bound of the given dimension
+        WI_NODISCARD LONG ubound(UINT dim = 1) const
+        {
+            static_assert(wistd::is_same<void, result>::value, "this method requires exceptions or failfast");
+            LONG result = 0;
+            ubound(dim, &result);
+            return result;
+        }
+        //! Return the total number of elements in the SAFEARRAY
+        WI_NODISCARD ULONG size() const
+        {
+            static_assert(wistd::is_same<void, result>::value, "this method requires exceptions or failfast");
+            ULONG result = 0;
+            size(&result);
+            return result;
+        }
+        //! Returns the number of elements in the given dimension
+        WI_NODISCARD ULONG dim_size(UINT dim = 1) const
+        {
+            static_assert(wistd::is_same<void, result>::value, "this method requires exceptions or failfast");
+            ULONG result = 0;
+            dim_size(dim, &result);
+            return result;
+        }
+        //! Returns a copy of the SAFEARRAY, including all individual elements
+        WI_NODISCARD unique_type create_copy() const
+        {
+            static_assert(wistd::is_same<void, result>::value, "this method requires exceptions or failfast");
+
+            auto result = unique_type{};
+            result.create_copy(storage_t::get());
+            return result;
+        }
+
+        //! Returns a safearraydata_t that provides direct access to this SAFEARRAY's contents.  See @ref safearraydata_t.
+        //! ~~~~
+        //! // Create a new safearray by copying a vector of wil::unique_bstr
+        //! unique_bstr_safearray copy_to_safearray(std::vector<wil::unique_bstr>& source)
+        //! {
+        //!     auto sa = wil::unique_bstr_safearray{source.size()};
+        //!     auto current = std::begin(source);
+        //!     for(BSTR& bstr : sa.access_data())
+        //!     {
+        //!         // Create a copy for the safearray to own
+        //!         bstr = ::SysAllocString((*current++).get());
+        //!     }
+        //!     return std::move(sa);
+        //! }
+        //! ~~~~
+        template<typename T = element_t, typename wistd::enable_if<is_type_set<T>::value, int>::type = 0>
+        WI_NODISCARD safearraydata_t<T> access_data() const
+        {
+            static_assert(wistd::is_same<void, result>::value, "this method requires exceptions or failfast");
+            return safearraydata_t<T>{ storage_t::get() };
+        }
+
+        //! Special Get Functions
+#if defined(__WIL_OLEAUTO_H_)

```

Comment:
This won't work for member functions. You are now introducing an include order dependency which should be avoided

Reply:
1) What won't work for member functions?  I don't understand that part of your comment.
2) I'm not introducing an include order dependency.  That dependency was already there.  People are told to include the system headers they're going to use first, and the WIL will customize itself based on the included headers.  I only want to include that functionality if WIL has supplied the types that I need.

Since (thanks to feedback on this PR) I am now in a separate header and am including the other headers I need, I will make sure to include `OleAuto.h` before including `resource.h`.  That way, if `resource.h` was included before `OleAuto.h` was included the first time, it will pick up these definitions the 2nd time around (assuming all the conditions are met).  Since I don't want to care about those conditions, I just want to add the functionality if they were met, testing for `__WIL_OLEAUTO_H_` will accomplish that.

TL;DR:  I'll include the headers I need in the right order and that will eliminate any order dependencies.